### PR TITLE
Share object literal types

### PR DESCRIPTION
**Pull request by ahejlsberg**
With this PR we share object literal type instances whenever possible. Previously we would always generate a new type object for every object literal expression. This optimization reduces the number of redundant type objects in codebases with large numbers of object literals of similar shape. For example, it reduces the check time and overall compile time for the Monaco project by 8% and 3% respectively.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -12810,6 +12815,13 @@ namespace ts {
                     undefined;
         }
 
+        function getContainingStatement(node: Node): Node {
```
Comment (sandersn) [inline src/compiler/checker.ts]: `findAncestor(node, isStatement)`

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -9003,7 +9006,7 @@ namespace ts {
                         //          breaking the intersection apart.
                         result = someTypeRelatedToType(<IntersectionType>source, target, /*reportErrors*/ false);
                     }
-                    if (!result && (source.flags & TypeFlags.StructuredOrTypeVariable || target.flags & TypeFlags.StructuredOrTypeVariable)) {
+                    if (!result && (source.flags & TypeFlags.StructuredOrTypeVariable || target.flags & TypeFlags.ObjectOrTypeVariable)) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: Why don't we cover unions and intersections here anymore? Actually, a better question is: Why did we handle them before? I thought recursiveTypeRelatedTo was for object types only. Was this a mistake? Should it have been only intersections, but not unions?

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -17528,6 +17574,8 @@ namespace ts {
             const secondAssignableToFirst = isTypeAssignableTo(type2, type1);
```
Comment (sandersn) [inline src/compiler/checker.ts]: aside: getBestChoiceType is a bad name. "Choice type" doesn't reflect that this is only used in `?:` and `||`. "Conditional result type" is better but would make for a clunky function name.

File: src/compiler/checker.ts [hunk 4]
```diff
@@ -13678,8 +13693,6 @@ namespace ts {
                     spread = getSpreadType(spread, createObjectLiteralType());
                 }
                 if (spread.flags & TypeFlags.Object) {
-                    // only set the symbol and flags if this is a (fresh) object type
```
Comment (sandersn) [inline src/compiler/checker.ts]: this comment should stay, although maybe it should explain that sometimes getSpreadType returns anyType (for example).

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -7834,7 +7835,9 @@ namespace ts {
                     members.set(leftProp.escapedName, getNonReadonlySymbol(leftProp));
                 }
             }
-            return createAnonymousType(undefined, members, emptyArray, emptyArray, stringIndexInfo, numberIndexInfo);
+            const result = createAnonymousType(undefined, members, emptyArray, emptyArray, stringIndexInfo, numberIndexInfo);
+            result.flags |= (left.flags | right.flags) & TypeFlags.PropagatingFlags;
```
Comment (sandersn) [inline src/compiler/checker.ts]: is this required as part of this change or is it drive-by fix up, a smarter way to propagate propagating flags?

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -13692,23 +13705,56 @@ namespace ts {
...
                     result.pattern = node;
                 }
-                if (!(result.flags & TypeFlags.Nullable)) {
-                    propagatedFlags |= (result.flags & TypeFlags.PropagatingFlags);
-                }
                 return result;
             }
         }
 
+        function isInferredTypeOfVariable(node: Node) {
```
Comment (sandersn) [inline src/compiler/checker.ts]: what about
```ts
declare function f<T>(t: T): T;
f({ a: 12 }).a
f({ a: 12 }).a/*1*/
```

Won't goto def at `/*1*/` go to the first object literal?

File: src/compiler/checker.ts [hunk 7]
```diff
@@ -13692,23 +13705,56 @@ namespace ts {
...
+            const parent = node.parent;
+            return isVariableLike(parent) && !parent.type && parent.initializer === node;
+        }
+
+        function propertyNameIsValidKey(symbol: Symbol) {
+            return (<string>symbol.escapedName).indexOf(":") < 0;
+        }
+
+        function getObjectLiteralType(flags: TypeFlags, symbol: Symbol, members: SymbolTable, properties: Symbol[]): ResolvedType {
+            const key = map(properties, p => p.escapedName + (p.flags & SymbolFlags.Optional ? ":?" : ":") + getTypeOfSymbol(p).id).join(",");
```
Comment (sandersn) [inline src/compiler/checker.ts]: calling `getTypeOfSymbol` here is a lot more eager than before. Probably not a problem?

**Review by sandersn — CHANGES_REQUESTED** [review]

File: tests/baselines/reference/assignmentCompatability_checking-apply-member-off-of-function-interface.errors.txt [hunk 8]
```diff
@@ -55,7 +54,6 @@ tests/cases/compiler/assignmentCompatability_checking-apply-member-off-of-functi
     fn({});
        ~~
 !!! error TS2345: Argument of type '{}' is not assignable to parameter of type 'Applicable'.
-!!! error TS2345:   Property 'apply' is missing in type '{}'.
```
Comment (sandersn) [inline tests/baselines/reference/assignmentCompatability_checking-apply-member-off-of-function-interface.errors.txt]: I'd rather not lose this line. Is it because we're caching `{}` now?

Comment (sandersn) [reply tests/baselines/reference/assignmentCompatability_checking-apply-member-off-of-function-interface.errors.txt]: Never mind, it's already reported above, so the improved caching dedupes it.

File: tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.symbols [hunk 9]
```diff
@@ -93,7 +93,7 @@ for (let {name: nameA = "noName" } of getRobots()) {
 >nameA : Symbol(nameA, Decl(sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.ts, 31, 10))
 }
 for (let {name: nameA = "noName" } of [{ name: "mower", skill: "mowing" }, { name: "trimmer", skill: "trimming" }]) {
->name : Symbol(name, Decl(sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.ts, 34, 40))
```
Comment (sandersn) [inline tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.symbols]: looks like isInferredTypeOfVariable misses the for-of case as well.

File: src/compiler/checker.ts (hunk 1 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts]: Will fix.

File: src/compiler/checker.ts (hunk 5 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts]: It's both cleanup and a necessary change because we only want to set the propagated flags when we are actually creating the type (as opposed to sharing an existing one).

File: src/compiler/checker.ts (hunk 2 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts]: We should exclude unions and intersections on the target side because the result will always be false. It always was false previously, but took longer to compute.

File: src/compiler/checker.ts (hunk 6 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts]: No, it will go to the second object literal. But if you then do a goto def on the line above, it also goes to the second object literal. You basically end up going to the first actual instantiation of the type produced by the compiler.

The PR currently shares all object literals types that don't originate from the right hand side of an initializer for a variable with no type annotation. We could exclude other cases (such as arguments to functions), but I don't think the current behavior is totally unreasonable.

File: src/compiler/checker.ts (hunk 7 above)
Comment (ahejlsberg) [reply src/compiler/checker.ts]: Not a problem, any symbol we touch here already had its `type` property explicitly initialized by the code that calls this function.

File: tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.symbols (hunk 9 above)
Comment (ahejlsberg) [reply tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.symbols]: Yeah, but again, not sure how much it matters.

File: src/compiler/checker.ts (hunk 6 above)
Comment (sandersn) [reply src/compiler/checker.ts]: Here's the worst case that I could come up with. Both `/*1*/` and `/*2*/` will jump to the same definition, I think.

1. I made meaningful names.
2. I split usages across multiple modules.
3. I made two 'different' types that are structurally identical.

(3) is the least realistic here, but this is confusing even if the types are the same. At scale, I think that this is a bug, but maybe not one worth fixing. I guess it boils down to how much excluding arguments hurts performance in our tests?


```ts
// @Filename: functional.ts
export function id<T>(t: T): T { return t; }

// @Filename: people.ts
import { id } from './functional'
export let bob = id({ name: 'Bob', age: 30 });
// @Filename: usage1.ts
import { bob } from './people'
bob.name/*2*/

// @Filename: cars.ts
export let mazda = id({ name: 'cx-5', age: 2 });
// @Filename: usage2.ts
import { mazda } from './cars'

mazda.name/*1*/
```

File: tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.symbols (hunk 9 above)
Comment (sandersn) [reply tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.symbols]: It probably matters less than the argument case, but it's the same kind of bug in a large code base.

File: tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPattern.symbols [hunk 10]
```diff
@@ -93,7 +93,7 @@ for (let {name: nameA } of getRobots()) {
 >nameA : Symbol(nameA, Decl(sourceMapValidationDestructuringForOfObjectBindingPattern.ts, 31, 10))
 }
 for (let {name: nameA } of [{ name: "mower", skill: "mowing" }, { name: "trimmer", skill: "trimming" }]) {
->name : Symbol(name, Decl(sourceMapValidationDestructuringForOfObjectBindingPattern.ts, 34, 29))
+>name : Symbol(name, Decl(sourceMapValidationDestructuringForOfObjectBindingPattern.ts, 16, 24))
```
Comment (mhegazy) [inline tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPattern.symbols]: this does not seem right, maybe exclude right hand of for..of?

File: tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.symbols (hunk 9 above)
Comment (mhegazy) [reply tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.symbols]: it matters for goto def usually. it is a bit disorienting to jump to hit F12 and go somewhere unrelated but similar.

Comment (mhegazy) [reply tests/baselines/reference/sourceMapValidationDestructuringForOfObjectBindingPatternDefaultValues.symbols]: also a bit unexpected for renames.
